Error checking with successful_step function

rusandris opened the pull request:
Hi! I've added `successful_step` as part of the DynamicalSystemsBase API. For now it seemed like a good first step to implement it for `ContinuousTimeDynamicalSystem` type (it uses already available `SciMLBase` functions), but it can be extended to others as well, depending how general we want it to be. 
An older version of this PR was discussed at https://github.com/JuliaDynamics/DynamicalSystemsBase.jl/pull/147 and had the purpose to have an error checking tool that can be used to stop trajectory, lyapunov etc. calculations in case stepping is unsuccessful.

Datseris reviewed (commented):
Thanks so much @rusandris I just realized I need this in another project as well!! Some comments:

The function should have a docstring. I suggest:
```
    successful_step(ds::DynamicalSystem) -> true/false
Return `true` if the last `step!` call to `ds` was successful, `false` otherwise.
```



---

Is it possible to disable diffeq warnings like `dt < dt` min, etc.? Then we can throw our own warnings if we want to.

---

The function should also be implemented for discrete time. For those systems, it should check:
```
any(x -> isinf(x) || isnan(x), current_state(ds))
```

Hope this helps!

Datseris commented:
Also, add the function to the API in the docstring of `DynamicalSystem`, and expand it to the docs as well. And increase version to 3.1.0 and add a new changelog entry !

rusandris commented:
I've added the default and discrete methods, please check if they are in the right place. Also added docstrings and hopefully expanded to docs as well.  

> Is it possible to disable diffeq warnings like dt < dt min, etc.? Then we can throw our own warnings if we want to.

Yes, there is [Supressor.jl ](https://github.com/JuliaIO/Suppressor.jl) but I don't know if that's a good idea :).

Datseris commented on src/core/dynamicalsystem_interface.jl:
```diff
@@ -241,6 +242,18 @@ but a developer may care.
+
+"""
+successful_step(ds::DynamicalSystem) = true
+successful_step(ds::DiscreteTimeDynamicalSystem) = all(x -> isfinite(x) , current_state(ds))
```
You forgot the `!isnan` check here!

rusandris replied:
Note that this is not `isinf`. I used `isfinite` which gives false for both `Inf`s and `NaN`s

Datseris replied:
ok perfect you are right!

Datseris commented on src/core/dynamicalsystem_interface.jl:
```diff
@@ -241,6 +242,18 @@ but a developer may care.
+for discrete time it checks if any variable is `Inf` or `NaN`.
+
+"""
+successful_step(ds::DynamicalSystem) = true
```
I think we should leave this undefined instead of putting `true` everywhere.

Datseris commented on src/core_systems/continuous_time_ode.jl:
```diff
@@ -143,6 +143,7 @@ initial_state(integ::DEIntegrator) = integ.sol.prob.u0
 current_state(integ::DEIntegrator) = integ.u
 current_time(integ::DEIntegrator) = integ.t
 initial_time(integ::DEIntegrator) = integ.sol.prob.tspan[1]
+successful_step(integ::DEIntegrator) = successful_retcode(check_error(integ))
```
For clarity, I'd do `using SciMLBase: successful_retcode, check_error` before this new function, so that we know where these names are coming from.

Datseris reviewed (changes requested):
Thanks a lot! There are two major things missing here:

1. There are no tests of this new functionality
2. This doesn't propagate to other dynamical sytems. You need to add the extension to parallel, tangent, and projected dynamical systems as well.

Datseris commented:
> Yes, there is [Supressor.jl ](https://github.com/JuliaIO/Suppressor.jl)but I don't know if that's a good idea :).

Nah, that is not the way to go. I was thinking whether there was a keyword argument one gives when initializing an ODEIntegrator. But I guess you are right, there is no reason to suppress this really.

Datseris commented on test/successful_step.jl:
```diff
@@ -0,0 +1,44 @@
+		for _ in 1:100
+			step!(sys)
+		end
+		@show sys
```
```suggestion
```

Datseris commented on src/core/dynamicalsystem_interface.jl:
```diff
@@ -241,6 +242,19 @@ but a developer may care.
+for discrete time it checks if any variable is `Inf` or `NaN`.
+
+"""
+successful_step(ds::DynamicalSystem) = nothing
```
```suggestion
successful_step(ds::DynamicalSystem) = errormsg(ds)
```

So that it correctly shows the error instead of being silent.

Datseris reviewed (approved):
This looks great to me! I'll let the tests run!

Datseris commented on src/core/dynamicalsystem_interface.jl:
```diff
@@ -241,6 +242,19 @@ but a developer may care.
+Return `true` if the last `step!` call to `ds` was successful, `false` otherwise.
+For continuous time systems this uses DifferentialEquations.jl error checking,
+for discrete time it checks if any variable is `Inf` or `NaN`.
+
```
```suggestion
```

Datseris commented on test/successful_step.jl:
you've added the test file but you have not actually included it in the tests!

Datseris commented on test/successful_step.jl:
```diff
@@ -0,0 +1,43 @@
+	#projected
+	projection = [1]
+	complete_state = [0.0]
+	pr = ProjectedDynamicalSystem(ode, projection, complete_state)
```
In this and all following creations of systems use `deepcopy(ode)`. Becuase otherwise you have a reference to the same system.

rusandris replied:
Right. I had that but deleted them for some reason. I add them back in

rusandris replied:
Btw can you explain the reason behind this? Why is the underlying `ds` passed by reference and not by value?

Datseris replied:
in DynamicalSystems 3.0 all instances of dynamical systems are integrators: mutable entities that are stepped in place. Julia always passes mutable entities by reference.

Datseris reviewed (changes requested):
ah the tests are not correct yet.

rusandris commented:
This being merged, do you need help with implementing this function into `trajectory`, `lyapunov` etc.?

Datseris commented:
You can add this to `lyapunov` and stuff, yes, but what is the reason to add this to `trajectory`? The function already behaves as it should when the step is not successful.

rusandris commented:
>  The function already behaves as it should when the step is not successful.

I used to and still get weird behaviour when for ex. `trajectory` meets and unstability. For continuous systems, the stepping doesn't stop, fills the REPL with the warning from SciML and it also gets stuck? Try this:

```
using DynamicalSystemsBase
using OrdinaryDiffEq

function unstable_sys(du,u,p,t)
           du[1] = u[1]
           du[2] = -u[2]
       return nothing
       end
u0 = [10.0, 10.0]
diffeq = (alg = Vern9(),abstol = 1e-9, reltol = 1e-9)

uns = CoupledODEs(unstable_sys, u0, nothing; diffeq)
trajectory(uns,200)

```
I suggest we use `successful_step` to fix this. Let me know what you think

Datseris commented:
Of course; we didn't actually use `successful_step` anywhere, so it makes sense that the behavior of `trajectory` didn't change. So yes I agree you should do a PR That uses `successful_step` in `trajectory`! I'm already using `succesful_step!` in other parts of the library!
